mpt_trie/src/trie_ops.rs

```diff
@@ -365,32 +396,36 @@ impl<T: PartialTrie> Node<T> {
 fn insert_into_trie_rec<N: PartialTrie>(
     node: &Node<N>,
     mut new_node: InsertEntry,
-) -> Option<WrappedNode<N>> {
+) -> TrieOpResult<Option<WrappedNode<N>>> {
     match node {
         Node::Empty => {
             trace!("Insert traversed Empty");
-            Some(create_node_from_insert_val(new_node.nibbles, new_node.v))
+            Ok(Some(create_node_from_insert_val(
+                new_node.nibbles,
+                new_node.v,
+            )))
         }
         Node::Hash(_) => {
             trace!("Insert traversed {:?}", node);
-            panic!(
-                "Found a `Hash` node during an insert in a `PartialTrie`! These should not be able to be traversed during an insert!"
-            )
+            Err(TrieOpError::HashNodeInsertError())
         }
         Node::Branch { children, value } => {
             if new_node.nibbles.count == 0 {
                 trace!("Insert traversed branch and placed value in node");
-                return Some(branch_from_insert_val(children.clone(), new_node.v));
+                return Ok(Some(branch_from_insert_val(children.clone(), new_node.v)?));
             }
 
             let nibble = new_node.nibbles.pop_next_nibble_front();
             trace!("Insert traversed Branch (nibble: {:x})", nibble);
 
-            insert_into_trie_rec(&children[nibble as usize], new_node).map(|updated_child| {
-                let mut updated_children = children.clone();
-                updated_children[nibble as usize] = updated_child;
-                branch(updated_children, value.clone())
-            })
+            insert_into_trie_rec(&children[nibble as usize], new_node)?.map_or(
+                Ok(None),
+                |updated_child| {
+                    let mut updated_children = children.clone();
+                    updated_children[nibble as usize] = updated_child;
+                    Ok(Some(branch(updated_children, value.clone())))
+                },
+            )
```

Comment:
I think the same thought with the wrapping the block in `Ok()` below. I'll let you call it either way here.